File: apps/desktop/src/lib/ipc/recovery.ts
// Mirror of `crates/oxyn-desktop/src/ipc/recovery.rs`, as executable schemas:
// both sides still change in the same commit, but a field renamed on one of
// them is now named at the boundary instead of surfacing as an `undefined`
// (ADR-0031).

import { Channel, isTauri } from "@tauri-apps/api/core"
import { z } from "zod"

import { call, guarded, Nothing } from "./client"
import { Environment, TransactionState } from "./types"

export const RecoveryStatus = z.object({
  /** Observed at startup: only then may the screen say Oxyn did not close normally. */
  abnormal: z.boolean(),
  /** History holds an execution the library marks `needsInspection`. */
  unresolvedWrite: z.boolean(),
})
export type RecoveryStatus = z.infer<typeof RecoveryStatus>

/** A console session whose transaction holds the exit (ADR-0043). */
export const ExitTransaction = z.object({
  session: z.string(),
  /** What `run_console` takes. Never shown: the name is. */
  connection: z.string(),
  connectionName: z.string(),
  environment: Environment,
  /** `open` or `unknown`: an idle session is not listed. */
  state: TransactionState,
})
export type ExitTransaction = z.infer<typeof ExitTransaction>

/** What an open transaction holds: the application's exit, or this window's close. */
export const ExitScope = z.enum(["application", "window"])
export type ExitScope = z.infer<typeof ExitScope>

export const ShutdownSignal = z.discriminatedUnion("type", [
  z.object({ type: z.literal("flushDrafts") }),
  z.object({
    type: z.literal("resolveTransactions"),
    /** This window's consoles only. */
    transactions: z.array(ExitTransaction),
    scope: ExitScope,
  }),
  z.object({ type: z.literal("exitCancelled") }),
])
export type ShutdownSignal = z.infer<typeof ShutdownSignal>

export const recovery = {
  status: () => call("recovery_status", RecoveryStatus),

  subscribeShutdown: (onSignal: (signal: ShutdownSignal) => void) => {
    // The channel registers a callback in Tauri's internals: constructing it
    // outside the webview throws before `call` can explain why.
    if (!isTauri()) return call("subscribe_shutdown", Nothing)
    const channel = new Channel<unknown>()
    channel.onmessage = guarded("subscribe_shutdown", ShutdownSignal, onSignal)
    return call("subscribe_shutdown", Nothing, { channel })
  },

  shutdownFlushed: () => call("shutdown_flushed", Nothing),

  /** The dialog of `resolveTransactions` or of a window close is shown: the exit waits for it. */
  shutdownAcknowledged: () => call("shutdown_acknowledged", Nothing),

  /** Abandons an exit held by a transaction, or this window's close; nothing was flushed. */
  cancelExit: () => call("cancel_exit", Nothing),

  /** The ordered exit of ⌘Q, asked by `File ▸ Exit`: takes nothing, chooses nothing. */
  requestExit: () => call("request_exit", Nothing),
}

File: crates/oxyn-desktop/src/ipc/recovery.rs
//! What crosses the boundary for recovery and shutdown
//! ([ADR-0021](../../../../docs/adr/0021-marqueur-d-arret.md)).

use serde::Serialize;

/// How the previous launch ended, as observed before anything was written.
#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecoveryStatus {
    /// The previous session was left open and silent. Only then may the
    /// recovery screen say Oxyn did not close normally.
    pub abnormal: bool,
    /// History holds an execution whose server state must be inspected — the
    /// library's `needs_inspection`. Only then does the screen warn about a
    /// write with an unknown outcome.
    ///
    /// Read at startup, like `abnormal`: the screen speaks of what the launch
    /// found. A write that turns ambiguous during this session is flagged in
    /// the library instead, and re-reading history here would be a second path
    /// to it beside the command bus (I-01).
    pub unresolved_write: bool,
}

/// What the backend asks of the webview while the window closes.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ShutdownSignal {
    /// Submit every draft still waiting for its typing pause, then answer
    /// `shutdown_flushed`.
    FlushDrafts,
    /// The exit is held by these transactions: answer `shutdown_acknowledged`
    /// at once, then let the user commit, roll back or cancel
    /// ([ADR-0043](../../../../docs/adr/0043-multi-fenetre.md)).
    #[serde(rename_all = "camelCase")]
    ResolveTransactions { transactions: Vec<ExitTransaction> },
    /// The exit was cancelled: close the dialog, nothing was flushed.
    ExitCancelled,
}

/// A console session whose transaction holds the exit.
///
/// Carries no SQL and no value: the connection's name and marking, and the
/// state the executor last observed.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExitTransaction {
    pub session: String,
    /// The id `run_console` takes. Never shown: the name is.
    pub connection: String,
    pub connection_name: String,
    pub environment: oxyn_core::Environment,
    /// `open` or `unknown`: an `idle` session is not listed.
    pub state: super::TransactionStateView,
}
